File: Arrays/intersection_array.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;
vector<int> intersection_array(vector<int>&a,vector<int>&b){
    vector<int>ans;
    int n1=a.size();
    int n2=b.size();
    int vis[n2]={0};
    for(int i=0;i<n1;i++){
        for(int j=0;j<n2;j++){
        if(a[i]==b[j] && vis[j]==0){
            ans.push_back(b[j]);
            vis[j]=1;
            break;
        }
        if(b[j]>a[i]) break;
        } 
    }
    return ans;
}

vector<int>intersection_array2(vector<int>&a,vector<int>&b){
    int n=a.size();
    int i=0,j=0;
    vector<int>ans;
    while(i<n && j<n){
        if(a[i]<b[j]) i++;
        else if(b[j]<a[i]) j++;
        else{
            ans.push_back(a[i]);
            i++;
            j++;
        }
    }
    return ans; 
}
int main(){
    int n1,n2;
    cout<<"Enter the size of the arr1 and arr2: ";
    cin>>n1>>n2;
    vector<int>arr1(n1),arr2(n2);
    cout<<"Enter the elements arr1: ";
    for(int i=0;i<n1;++i){
        cin>>arr1[i];
    }
    cout<<"Enter the elements arr2: ";
    for(int i=0;i<n2;++i){
        cin>>arr2[i];
    }
    vector<int>unionarr=intersection_array(arr1,arr2);
    for(auto it: unionarr) cout<<it<<" ";

    return 0;
}

File: java/ArrayQ/IntersectionArray.java
package ArrayQ;

import java.util.*;

public class IntersectionArray {

    // Method 1: Brute force using visited array
    public static List<Integer> intersectionArray(int[] a, int[] b) {
        List<Integer> ans = new ArrayList<>();

        int n1 = a.length;
        int n2 = b.length;
        int[] vis = new int[n2]; // visited array

        for (int i = 0; i < n1; i++) {
            for (int j = 0; j < n2; j++) {

                if (a[i] == b[j] && vis[j] == 0) {
                    ans.add(b[j]);
                    vis[j] = 1;
                    break;
                }

                if (b[j] > a[i]) break; // optimization for sorted arrays
            }
        }
        return ans;
    }

    // Method 2: Optimal two-pointer approach (sorted arrays)
    public static List<Integer> intersectionArray2(int[] a, int[] b) {
        List<Integer> ans = new ArrayList<>();

        int i = 0, j = 0;
        int n1 = a.length, n2 = b.length;

        while (i < n1 && j < n2) {
            if (a[i] < b[j]) {
                i++;
            } else if (b[j] < a[i]) {
                j++;
            } else {
                ans.add(a[i]);
                i++;
                j++;
            }
        }
        return ans;
    }

    // Driver code
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter the size of arr1 and arr2: ");
        int n1 = sc.nextInt();
        int n2 = sc.nextInt();

        int[] arr1 = new int[n1];
        int[] arr2 = new int[n2];

        System.out.print("Enter the elements arr1: ");
        for (int i = 0; i < n1; i++) {
            arr1[i] = sc.nextInt();
        }

        System.out.print("Enter the elements arr2: ");
        for (int i = 0; i < n2; i++) {
            arr2[i] = sc.nextInt();
        }

        List<Integer> intersection = intersectionArray(arr1, arr2);

        for (int x : intersection) {
            System.out.print(x + " ");
        }

        sc.close();
    }
}
